### PR TITLE
Stage2

### PR DESCRIPTION
**Pull request by VeronikaVasilchikova**
Date: 2020/03/19

Here is my implementation of stage2 according to requirements of https://hackmd.io/@listopad/rk_RV6fez

I'll be glad to get your feedback and comments.

### CODE REVIEW

File: sources/views/datatableActivities.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+					onClick: {
+						"wxi-trash": (e, id) => {
+							webix.confirm({
+								title: "Remove this note",
+								ok: "Yes",
+								cancel: "No",
+								text: "Are you sure you want to remove this note?"
+							})
+								.then(() => {
+									webix.confirm({
```
Comment (ByTiger) [inline sources/views/datatableActivities.js]: The chain of promises can be simplified.
=> promise(...).then(return promise(...)).then(return promise(...)).then(...)

File: sources/views/formForContact.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+	urlChange() {
+		webix.promise.all([
+			statuses.waitData,
+			contacts.waitData
+		]).then(() => {
+			const id = this.getParam("id", true);
+			const item = contacts.getItem(id);
+			if (this.getParam("value") === "edit") {
```
Comment (HelgaListopad) [inline sources/views/formForContact.js]: Template "contactPhoto" is not a part of the form (and can't be), so you need to set the current photo to it here.

File: sources/views/formForContact.js [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+	closeForm() {
+		this.form.clear();
+		this.form.clearValidation();
+		this.show("./details");
+	}
+
+	addOrEdit() {
+		if (this.form.validate()) {
+			const values = this.form.getValues();
```
Comment (HelgaListopad) [inline sources/views/formForContact.js]: Photo is not saved with other values

File: sources/views/details.js [hunk 4]
```diff
@@ -1,37 +1,35 @@
...
-					</div>
-				`
-				}
-			]
+			localId: "contactsTemplate",
+			template: `
+			<h2>#FirstName# #LastName#</h2>
+			<div class='details'>
+				<div>
+					<image class="contactphoto" src="data/photo/contact_photo.jpg" />
```
Comment (HelgaListopad) [inline sources/views/details.js]: Use `obj.Photo` here. If the user has no photo, use the current placeholder.

File: sources/views/contacts.js [hunk 5]
```diff
@@ -14,22 +14,33 @@ export default class ContactsView extends JetView {
...
+								{
+									width: 300,
+									view: "list",
+									localId: "listOfContacts",
+									scroll: "y",
+									select: true,
+									type: {
+										template: obj => `
+											<div class="contact">
+												<image class="contactphoto" src="data/photo/contact_photo.jpg" />
```
Comment (HelgaListopad) [inline sources/views/contacts.js]: Use `obj.Photo` here. If the user has no photo, use the current placeholder.

File: sources/views/formForContact.js [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+			contacts.waitSave(() => {
+				if (values && values.id) {
+					contacts.updateItem(values.id, values);
+					webix.message({type: "success", text: "Contact was updated successfully!"});
+				}
+				else {
+					contacts.add(values, 0);
```
Comment (HelgaListopad) [inline sources/views/formForContact.js]: Newly added contact should be selected after saving - a good task for `waitSave` promise

File: sources/views/details.js [hunk 7]
```diff
@@ -1,37 +1,35 @@
...
+					<div class="status">Status #newStatusID#</div>
+				</div>
+				<div>
+					<i class="fas fa-envelope"></i>#Email#<br><br>
+					<i class="fab fa-skype"></i>#Skype#<br><br>
+					<i class="fas fa-pencil-ruler"></i>#Job#<br><br>
+					<i class="fas fa-briefcase"></i>#Company#
+				</div>
+				<div>
+					<span class='webix_icon wxi-calendar'></span>#Birthday#<br><br>
```
Comment (HelgaListopad) [inline sources/views/details.js]: `Birthday` is a Date, so some date-to-str formating is required here

File: sources/views/details.js [hunk 8]
```diff
@@ -82,9 +104,48 @@ export default class DetailsView extends JetView {
 				const contact = webix.copy(contacts.getItem(id));
 				if (contact.StatusID) {
 					contact.newStatusID = statuses.getItem(contact.StatusID).Value || "";
+					contact.Birthday = webix.i18n.longDateFormatStr(contact.Birthday);
```
Comment (HelgaListopad) [inline sources/views/details.js]: Seems like it does not work for newly added items. Better do that in the template.

File: sources/views/details.js (hunk 7 above)
Comment (VeronikaVasilchikova) [reply sources/views/details.js]: I fixed it

File: sources/views/details.js (hunk 8 above)
Comment (VeronikaVasilchikova) [reply sources/views/details.js]: I fixed it

File: sources/views/contacts.js (hunk 5 above)
Comment (VeronikaVasilchikova) [reply sources/views/contacts.js]: I fixed it

File: sources/views/details.js (hunk 4 above)
Comment (VeronikaVasilchikova) [reply sources/views/details.js]: I fixed it

File: sources/views/datatableFiles.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import {fileStorage} from "../models/filestorage";
+import {contacts} from "../models/contacts";
+
+export default class DatatableFilesView extends JetView {
+	config() {
+		return {
+			rows: [
+				{
+					view: "datatable",
+					id: "datatableFile",
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableFiles.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+import {activities} from "../models/activities";
+import {activitytypes} from "../models/activitytypes";
+
+export default class DatatableActivitiesView extends JetView {
+	config() {
+		return {
+			rows: [
+				{
+					view: "datatable",
+					id: "datatableAct",
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/formForContact.js (hunk 2 above)
Comment (VeronikaVasilchikova) [reply sources/views/formForContact.js]: I fixed it

File: sources/views/formForContact.js (hunk 3 above)
Comment (VeronikaVasilchikova) [reply sources/views/formForContact.js]: I fixed it

File: sources/views/details.js [hunk 11]
```diff
@@ -82,9 +104,48 @@ export default class DetailsView extends JetView {
...
+			})
+				.then(() => {
+					webix.confirm({
+						title: "Warning!",
+						type: "confirm-warning",
+						text: "You are about to agree. Are you sure?"
+					})
+						.then(() => {
+							contacts.remove(id);
+							const idVal = Number(id);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/details.js]: I fixed it

File: sources/views/datatableFiles.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	init() {
+		this.$$("datatableFile").sync(fileStorage);
+	}
+
+	urlChange() {
+		const id = this.getParam("id", true);
+		this.$$("datatableFile").sync(fileStorage, () => {
+			this.$$("datatableFile").filter(item => item.ContactID === id);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableFiles.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+	addItem() {
+		const idForName = this.getParam("id", true);
+		this._jetPopupForm.showPopupForm("", idForName);
+	}
+
+	urlChange() {
+		activities.waitData.then(() => {
+			const id = this.getParam("id", true);
+			const idVal = Number(id);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+		);
+	}
+
+	init() {
+		this._jetPopupForm = this.ui(PopupFormView);
+		this.table = this.$$("datatableAct");
+
+		activities.waitData.then(() => {
+			const id = this.getParam("id", true);
+			const idVal = Number(id);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+				}
+			]
+		};
+	}
+
+	filterActivities(id) {
+		this.table = this.$$("datatableAct");
+		this.table.clearAll();
+		this.table.parse(
+			activities.find(item => item.ContactID === id)
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/formForContact.js (hunk 6 above)
Comment (VeronikaVasilchikova) [reply sources/views/formForContact.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+		activities.waitData.then(() => {
+			const id = this.getParam("id", true);
+			const idVal = Number(id);
+			this.filterActivities(idVal);
+		});
+
+		this.on(this.app, "onClickSave", (values) => {
+			if (values) {
+				this.table.parse(values);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+			}
+			this.table.refresh();
+		});
+	}
+
+	editItem(id) {
+		if (id) {
+			this._jetPopupForm.showPopupForm(id);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/datatableActivities.js (hunk 1 above)
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

File: sources/views/datatableActivities.js [hunk 18]
```diff
@@ -89,56 +87,40 @@ export default class DatatableActivitiesView extends JetView {
...
 
-	addItem() {
+	editOrAddItem(id) {
 		const idForName = this.getParam("id", true);
-		this._jetPopupForm.showPopupForm("", idForName);
+		if (id && idForName) {
+			this._jetPopupForm.showPopupForm(id, idForName);
+		}
+		if (!id && idForName) {
+			this._jetPopupForm.showPopupForm("", idForName);
```
[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

[the remark this answers is not in the document]
Comment (HelgaListopad) [reply sources/views/datatableActivities.js]: In addition, reduce the code here in datatableActivities.js ( where the comment was left (: )
```
		if (id && idForName) {
			this._jetPopupForm.showPopupForm(id, idForName);
		}
		if (!id && idForName) {
			this._jetPopupForm.showPopupForm("", idForName);
```
to 
```
		if (idForName) {
			this._jetPopupForm.showPopupForm(id, idForName);
		}
```
Why you don;t need complex `if-if`: 
- `showPopupForm` if called anyway (regardless of whether `id` exists at this point)
- `id` is checked in showPopupForm, so `undefined` will be handled there - you don't need to worry about it here.

[the remark this answers is not in the document]
Comment (VeronikaVasilchikova) [reply sources/views/datatableActivities.js]: I fixed it

**Pull request merged**